### PR TITLE
Flash Transformers modeling backend support

### PR DESCRIPTION
**Pull request by Cyrilvallez**
# What does this PR do?

Supersedes https://github.com/huggingface/text-generation-inference/pull/2858 for convenience (CIs) as requested (I cannot just change the origin branch from the old PR)

### CODE REVIEW

File: server/text_generation_server/utils/logits_process.py [hunk 1]
```diff
@@ -219,7 +218,7 @@ def filter(self, indices):
         return None
 
 
-class HeterogeneousTopPLogitsWarper(LogitsWarper):
+class HeterogeneousTopPLogitsWarper(LogitsProcessor):
```
Comment (Narsil) [inline server/text_generation_server/utils/logits_process.py:221]: Why those changes ?

It looks like these are Warpers, not Processors.... No ?

File: server/text_generation_server/models/transformers_flash_causal_lm.py [hunk 2]
```diff
@@ -0,0 +1,563 @@
...
+        kv_cache: List[KVCache],
+        block_tables: torch.Tensor,
+        slots: torch.Tensor,
+        seqlen: Seqlen,
+        max_s: int,
+        lm_head_indices: Optional[torch.Tensor],
+    ):
+        # Transformers does not support None as a default
+        if lm_head_indices is None:
+            lm_head_indices = 0
```
Comment (Narsil) [inline server/text_generation_server/models/transformers_flash_causal_lm.py:258]: But that's wrong isn´t  it ?

`0` would mean use index `0` which is probably wrong.
How does `transformers` use that value anyway ?

File: server/text_generation_server/models/transformers_flash_causal_lm.py [hunk 3]
```diff
@@ -0,0 +1,563 @@
...
+            block_tables=block_tables,
+            slots=slots,
+            seqlen=seqlen,
+            max_s=max_s,
+            kv_head_mapping=self.kv_head_mapping,
+        ).logits.squeeze(dim=0)
+        return logits
+
+
+    def forward(
```
Comment (Narsil) [inline server/text_generation_server/models/transformers_flash_causal_lm.py:278]: To save rewriting all of this, can we maybe monkey path `model.forward`.

It's not great, but it still feels better than this (which is duplicating quite complex logic that we're likely to forget to update).


in the `__init__`:  `self.model.forward = monkey_patch_forward` (that does the simple unsqueeze, squeeze).

Wdyt ?

The other option would be to introduce an indirection in `FlashCausalLM` by having an inner method that captures that `model.forward`call and we override it (I'm not a fan either to add such level of indirection, it's just as ugly as the monkey patching in my book, because it's still impossible for the reader to know that it's being modified somewhere else).

File: server/text_generation_server/utils/logits_process.py (hunk 1 above)
Comment (Cyrilvallez) [reply server/text_generation_server/utils/logits_process.py:221]: Recent Transformers version deprecated `LogitsWarper` in favor of `LogitsProcessor`. They have the same logic/interface

Comment (Cyrilvallez) [reply server/text_generation_server/utils/logits_process.py:221]: `LogitsWarper` has been completely removed, thus it needs to change here

File: server/text_generation_server/models/transformers_flash_causal_lm.py (hunk 2 above)
Comment (Cyrilvallez) [reply server/text_generation_server/models/transformers_flash_causal_lm.py:258]: We slice from `-num_logits_to_keep:`, so 0 means "keep all logits"

File: server/text_generation_server/models/transformers_flash_causal_lm.py (hunk 3 above)
Comment (Cyrilvallez) [reply server/text_generation_server/models/transformers_flash_causal_lm.py:278]: Yes I thought about it but was not sure you guys would like it 🙃 I'll update with the monkey patch

File: server/text_generation_server/models/transformers_flash_causal_lm.py (hunk 2 above)
Comment (Narsil) [reply server/text_generation_server/models/transformers_flash_causal_lm.py:250]: Then it's still wrong and we cannot use that.

TGI does dynamic batching, so the amount of logits to keep is just not enough information.


We can keep any amount of indices (or just the last ones, or whatever). Depending of if we need the logprobs, or however sliced the various queries are interlaced.

Maybe let's revert to doing this ourselves manually no ?

**Review by Narsil — APPROVED** [review]
LGTM. Thanks for this !

**Review by Narsil — APPROVED** [review]
